### PR TITLE
Allow setting case on Xrefs

### PR DESCRIPTION
**Pull request by PiotrKozlowski**
@aiwenar Edit: Handled. ~~`xmlnsCmlnle="http://katalysteducation.org/cmlnle/1.0"` is transformed to `xmlns-cmlnle="http://katalysteducation.org/cmlnle/1.0"`~~

### CODE REVIEW

File: src/cnxml/deserialize.js [hunk 1]
```diff
@@ -177,13 +177,14 @@ function emphasis(el) {
...
     const url = el.getAttribute('url') || null
+    const declension = el.getAttribute('cmlnle-case') || null
 
     if (target) {
         return {
             object: 'inline',
             type: 'xref',
             isVoid: true,
-            data: { target },
+            data: { target, declension },
```
Comment (aiwenar) [inline src/cnxml/deserialize.js]: This should be called _case_; _declension_ in English is not the same as a _grammatical case_.

> **declension**
>
> 1. A falling off, decay or descent.
> 2. (grammar) The act of declining a word; the act of listing the inflections of a noun, pronoun or adjective in order.
> 3. (grammar) A way of categorizing nouns, pronouns, or adjectives according to the inflections they receive. 
>
> (from [Wiktionary](https://en.wiktionary.org/wiki/declension))

> **(grammatical) case**
>
> _[..]_
>
> 6. (grammar) A specific inflection of a word depending on its function in the sentence.
> 
> and
>
> 1. (grammar) A mode of inflection of a word dependent on its use, especially the syntactic function in a phrase. 
>
> (from [Wiktionary](https://en.wiktionary.org/wiki/case#Noun), [Wiktionary](https://en.wiktionary.org/wiki/grammatical_case#English))

File: src/cnxml/serialize.js [hunk 2]
```diff
@@ -75,6 +75,14 @@ const BLOCK_TAGS = {
 }
 
 
+/**
+ * List of valid declensions for xrefs.
+ * 
+ * @see xref
+ */
+const DECLENSIONS = ['genitive', 'dative', 'accusative', 'instrumental', 'locative', 'vocative']
```
Comment (aiwenar) [inline src/cnxml/serialize.js]: You shouldn't limit it to just cases used in Polish. If you want a list of all existing grammatical cases (at least according to Wikipedia) I should still have it somewhere.

File: src/cnxml/serialize.js [hunk 3]
```diff
@@ -132,7 +140,22 @@ function list(obj, children) {
...
+    let attrs = {
+        'target-id': obj.data.get('target')
+    }
+
+    const declension = obj.data.get('declension')
+    const cmlnleCase = declension ?
+        DECLENSIONS.some(dec => dec === declension) ?
+            declension
+            : 'INVALID'
+        : null
```
Comment (aiwenar) [inline src/cnxml/serialize.js]: Why do you have this check here? You are guaranteed that the document you are serializing is correct according to its schema, and if you want to limit values of `cmlnle:case` then you should do it in schema.

File: src/plugins/xref/schema.js [hunk 4]
```diff
@@ -17,7 +17,8 @@ export default {
             // TODO: better data validation
             data: {
                 document: value => value == null || typeof value === 'string',
-                target: Boolean,
+                target: t => t === null || typeof t === 'string',
+                declension: d => d === null || d === undefined || typeof d === 'string',
```
Comment (aiwenar) [inline src/plugins/xref/schema.js]: 1. `target` must be defined, so this check is wrong
2. rather than testing for `null` and `undefined` separately, check just for `d == null`

File: src/plugins/xref/queries.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+  return inline.type === 'xref' ? inline : null
+}
+
+export function getActiveLink(editor, value) {
+  const inline = value.startInline
+  if (!inline) return null
+
+  return inline.type === 'link' ? inline : null
+}
```
Comment (aiwenar) [inline src/plugins/xref/queries.js]: Given how trivial those queries are, I'd rather not have them as separate functions.

File: src/plugins/xref/schema.js (hunk 4 above)
Comment (PiotrKozlowski) [reply src/plugins/xref/schema.js]: Not all xrefs will have `data.declension` so `undefined` is needes as well as `null`

File: src/cnxml/serialize.js (hunk 2 above)
Comment (PiotrKozlowski) [reply src/cnxml/serialize.js]: Sure, please send me if you find it or feel free to add it

Comment (PiotrKozlowski) [reply src/cnxml/serialize.js]: But wait, is cnxml supporting all those cases?

File: src/cnxml/serialize.js (hunk 3 above)
Comment (PiotrKozlowski) [reply src/cnxml/serialize.js]: I was thinking this way:
schema should allow any string as `case`, but when serializing to cnxml we have to be sure that it is supported.

File: src/plugins/xref/schema.js (hunk 4 above)
Comment (PiotrKozlowski) [reply src/plugins/xref/schema.js]: `target: typeof t === 'string'` will be ok then?

File: src/cnxml/serialize.js (hunk 2 above)
Comment (aiwenar) [reply src/cnxml/serialize.js]: This is our extension to CNXML, and there isn't even any document documenting it :) But the idea was that you can set any valid case.

File: src/cnxml/serialize.js (hunk 3 above)
Comment (aiwenar) [reply src/cnxml/serialize.js]: This makes no sense. Why would you allow invalid values anywhere?

File: src/plugins/xref/schema.js (hunk 4 above)
Comment (aiwenar) [reply src/plugins/xref/schema.js]: 1. Note that I proposed `d == null`, with only two equals signs. This is the same as checking with three equal signs for `d === null || d === undefined`

2. Technically in Slate keys need not be strings (or at least I think so), that's why I was only checking for their presence and not type

File: src/cnxml/serialize.js (hunk 3 above)
Comment (PiotrKozlowski) [reply src/cnxml/serialize.js]: In case when we want to serialize to another format which may be supporting those cases

File: src/plugins/xref/schema.js (hunk 4 above)
Comment (PiotrKozlowski) [reply src/plugins/xref/schema.js]: `d === null || d === undefined` - this is exactly why I've like to separate those checks, but it's your repo so I'll adjust

Comment (aiwenar) [reply src/plugins/xref/schema.js]: In general I agree that it's better to use explicit checks, but here we have so many checks that it just becomes hard to read.

File: src/cnxml/deserialize.js [hunk 6]
```diff
@@ -177,13 +177,14 @@ function emphasis(el) {
 function xref(el) {
     const target = el.getAttribute('target-id') || null
     const url = el.getAttribute('url') || null
+    const cmlnleCase = el.getAttribute('cmlnle-case') || null
```
Comment (aiwenar) [inline src/cnxml/deserialize.js]: This won't work, you need to use [`getAttributeNS`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getAttributeNS),

File: src/plugins/xref/schema.js [hunk 7]
```diff
@@ -18,6 +26,7 @@ export default {
             data: {
                 document: value => value == null || typeof value === 'string',
                 target: Boolean,
+                case: c => c == null || DECLENSIONS.some(d => d === c),
```
[the remark this answers is not in the document]
Comment (PiotrKozlowski) [reply src/plugins/xref/schema.js]: :man_facepalming: I was lately using `some()` ins other project and I've totally forgot about this

File: src/cnxml/deserialize.js (hunk 6 above)
Comment (PiotrKozlowski) [reply src/cnxml/deserialize.js]: There was type `-` instead of `:`
`getAttributeNS('cmlnle', 'case)` was not working but `getAttribute('cmlnle:case')` is ok.

Comment (aiwenar) [reply src/cnxml/deserialize.js]: `getAttribute('cmlnle:case')` also doesn't work:

```js
const document = `<?xml version="1.0" encoding="utf-8"?>
<document xmlns="http://cnx.rice.edu/cnxml" xmlns:cmlnle="http://katalysteducation.org/cmlnle/1.0" xmlns:n1="http://katalysteducation.org/cmlnle/1.0">
    <test cmlnle:case="nominative" />
    <test n1:case="accusative" />
</document>`

const tree = new DOMParser().parseFromString(d, 'application/xml')

console.log(t.children[0].children[0].getAttribute('cmlnle:case'))
// => nominative
console.log(t.children[0].children[1].getAttribute('cmlnle:case'))
// => null
```

The reason that `getAttributeNS` didn't work for you is because you used it wrong. `cmlnle` is _not_ a namespace name, its a namespace _prefix_, which the user is free to choose (as I did here, using both `cmlnle` and `n1` to refer to the same namespace). The namespace is actually `http://katalysteducation.org/cmlnle/1.0`. Continuing with example:

```js
console.log(t.children[0].children[0].getAttributeNS('http://katalysteducation.org/cmlnle/1.0', 'case'))
// => nominative
console.log(t.children[0].children[1].getAttributeNS('http://katalysteducation.org/cmlnle/1.0', 'case'))
// => accusative
```

**Comment by aiwenar** [comment]
I've just realised that transformation from `nsprefixAttrname` to `nsprefix:attrname` is not part of React or JSX but was implemented by me in [src/cnxml/xml.js](https://github.com/katalysteducation/cnx-designer/blob/master/src/cnxml/xml.js). All you need to do to make it work is add `cmlnle: 'http://katalysteducation.org/cmlnle/1.0',` to `NAMESPACES` in that file.

File: src/cnxml/deserialize.js (hunk 6 above)
Comment (PiotrKozlowski) [reply src/cnxml/deserialize.js]: Strange, I've tested on live editor and `getAttribute('cmlnle:case')` is working.

Comment (aiwenar) [reply src/cnxml/deserialize.js]: Apparently it does when you use the same namespace prefix as is used in the document, but it won't work if you use a different one, as I've shown.

**Comment by PiotrKozlowski** [comment]
@aiwenar Should we add something more here?

File: src/cnxml/index.js [hunk 8]
```diff
@@ -30,7 +30,7 @@ const serializer = new Html({
 export default {
     deserialize: (...args) => serializer.deserialize(...args),
 
-    serialize(value) {
+    serialize(value) {
```
Comment (aiwenar) [inline src/cnxml/index.js]: In the future strip such changes before sending a PR.

**Comment by aiwenar** [comment]
Also when possible try to make PR history linear, having multiple merge commits within a feature branch makes it hard to follow.

File: src/cnxml/index.js (hunk 8 above)
Comment (PiotrKozlowski) [reply src/cnxml/index.js]: How can I do this?
